### ISSUE TITLE
Gradients are None

### ISSUE THREAD

**Issue by stokhos** [tensorflow/probability#1165]
```python
loc_1 = tf.Variable([38.6], trainable=True)
logscale_1 = tf.Variable([2.3], trainable=True)

loc_2 = tf.Variable([6.6], trainable=True)
logscale_2 = tf.Variable([3.4], trainable=True)
         
a = tf.Variable([3.3], trainable=True)
b = tf.Variable([6.0], trainable=True)
joint = tfd.JointDistributionSequential([tfd.Normal(loc=loc_1, scale=tf.exp(logscale_1)),
                                         tfd.Normal(loc=loc_2, scale=tf.exp(logscale_2)),
                                         lambda b1, b0: tfd.Independent(
                                             tfd.Normal(
                                             loc=b0 + b1*3,
                                             scale=a/b),
                                             reinterpreted_batch_ndims=1
                                             ),])

with tf.GradientTape() as tape:
    ll = joint.log_prob(3,5,6)

tape.gradient(ll, [loc_1, loc_2,  logscale_1, logscale_2]
# gradients wrt logscale_1, logscale_2 are None

```

```python
rates_v = [tf.Variable(i+3.4, 
                       trainable=True,
                       name="mu:{}".format(i), 
                       constraint=lambda x: tf.clip(0.000001, np.infty)) for i in range(5)] 

possions_dist_v = tfd.Poisson(rate=rate_v)
joint_dist_v = tfd.JointDistributionSequential([tfd.Independent(possions_dist_v,0)]) 

with tf.GradientTape() as tape:       
    loss = -2*joint_dist_v.log_prob(samples[0])

tape.gradient(loss, rates_v)
# all None 
```
I thought I could get the gradients in these two cases. Did I make some mistakes?

**Comment by maxkaustav** [comment]
You cannot calculate in a distribution. Moreover, I will look into this to provide u an alternative.

**Comment by stokhos** [comment]
@maxkaustav I'm confused, what do you mean by calculate in a distribution?

**Comment by stokhos** [comment]
```python
loc_1 = tf.Variable(10.,name='loc_1'),
loc_2 = tf.Variable(20.,name='loc_2'),
scale_1 = tf.Variable(0.9, name='logscale_1') 
scale_2 = tf.Variable(0.9, name='logscale_2') 

joint=tfd.JointDistributionSequential([
                                 tfd.Normal(loc=loc_1,
                                            scale=tfp.util.TransformedVariable(scale_1,
                                                                               tfp.bijectors.Exp(),name='scale_1')),
                                 tfd.Normal(loc=loc_2,
                                            scale=tfp.util.TransformedVariable(scale_2, 
                                                                               tfp.bijectors.Exp(),name='scale_2')),
                                 lambda a,b :(
                                     tfd.Independent(tfd.Poisson(rate=tfp.util.TransformedVariable(a, tfp.bijectors.Exp()) * tfp.util.TransformedVariable(b, tfp.bijectors.Exp())), 0))
])

with tf.GradientTape() as tape:
    ll = joint.log_prob([10,30,[13]])

tape.gradient(ll, [loc_1,loc_2,scale_1,scale_2])
# [(None,), (None,), None, None]
```
Where there are () outside first two gradients?

**Comment by maxkaustav** [comment]
```
with tf.GradientTape(watch_accessed_variables=False) as tape:
      tf.nest. map_structure(tape.watch , model.trainable_variables)
grads = tape.gradient (loss , model.trainable_variables)
optimizer.apply_gradients(zip(tf.nest.flatten(grads),tf.nest.flatten(model.trainable_variables)))
```

Use this to compute gradients on distribution layers

**Comment by stokhos** [comment]
@maxkaustav I'm not sure how is your method different from mine. Can you be more specific?

In my opinion
your code
```python
with tf.GradientTape(watch_accessed_variables=False) as tape:
      tf.nest. map_structure(tape.watch , model.trainable_variables)
```
is equivlent to my code
```python
with tf.GradientTape() as tape:
```

**Comment by maxkaustav** [comment]
Like all TensorFlow distributions, gradients of log_prob and sample (for reparameterizeable distributions) with respect to any model parameters are available by reverse-mode automatic differentiation

I have shown a simpler version `tf.math.minimize`
